docs/csharp/language-reference/operators/null-parameter-check.md

```diff
@@ -0,0 +1,81 @@
+ms.date: 03/30/2022
+helpviewer_keywords:
+  - "null-forgiving operator [C#]"
+  - "!! operator [C#]"
+---
+# !! (null validation) operator (C# reference)
+
+The `!!` operator, introduced in C# 11, provides null validation parameter syntax. Adding `!!` to a parameter declaration instructs the compiler to add a runtime check for that parameter. For example:
+
+``` csharp
+void Method(string name!!)
+{
+    // ...
+}
+```
+
+generates code similar to the following example:
+
+``` csharp
+void Method(string name) 
+{
+    if (name is null)
+    {
+        throw new ArgumentNullException(nameof(name));
+    }
+    // ...
+}
+```
+
+Multiple annotated parameters are checked in the order declared on the method. For example, the following code:
+
+``` csharp
+void Method(string firstName!!, string lastName!!, string? nickName, string message)
+{
+    // ...
+}
+```
+
+generates code similar to the following example:
+
+```csharp
+void Method(string firstName, string lastName, string nickName, string message) 
+{
+    if (firstName is null)
+    {
+        throw new ArgumentNullException(nameof(firstName));
+    }
+    if (lastName is null)
+    {
+        throw new ArgumentNullException(nameof(lastName));
+    }
+    // ...
+}
+```
+
+It's intended for library authors to provide runtime checks even when APIs have been annotated for nullable reference types. Some libraries support projects that aren't nullable enabled. Some projects support multiple versions including some that predate nullable reference types. These checks can simplify the necessary validation. You should only add the `!!` operator on those parameters that require a null-check for safety. In the preceding example, the `message` parameter doesn't have a runtime null check.
+
+There are a few rules that govern where you can add the `!!` operator on a parameter declaration:
+
+- The `!!` operator directs the compiler to add runtime behavior. It can't be applied to a declaration without an implementation. You can't add `!!` to the parameter of an abstract member, an interface method without an implementation, or a parameter on a delegate type declaration.
+- The parameter type must be a type that can be compared to `null`. For example, the parameter can't be a non-nullable value type. If the parameter is a type parameter, it can't be constrained to be a non-nullable type (for example the `struct`, `notnull`, and `unmanaged` constraints are disallowed.)
+
+In almost all cases, the code the compiler generates for `!!` on an argument is consistent with the preceding example. However, the null check is inserted before any code you write in the method. The compiler generated null-check in constructors appears before the compiler generated calls to field initializers and before the call to any other constructor or base constructor. You can't catch the <xref:System.ArgumentNullException?displayProperty=nameWithType> thrown by the null check added by the compiler.
```

Comment:
```suggestion
In almost all cases, the code the compiler generates for `!!` on an argument is consistent with the preceding example. However, the null check is inserted before any code you write in the method. The compiler generated null-check in constructors appears before the compiler generated calls to field initializers and before the call to any other constructor or base constructor.

> [!IMPORTANT]
> You can't catch the <xref:System.ArgumentNullException?displayProperty=nameWithType> thrown by the null check added by the compiler.
```

This needs to stand out.

Reply:
Hmmm. I was almost tempted to delete this sentence. I think it would be writing nonsensical code.  Consider what the compiler generates:

```csharp
void M(string s)
{
    if (s is null)
    {
        throw new NullArgumentException(nameof(s));
    }
    // code....
}
```

Why would anyone write the following:

```csharp
void M(string s)
{
    try {
    if (s is null)
    {
        throw new NullArgumentException(nameof(s));
    }
    // code....
    } catch (NullArgumentException e)
    {
        // Why would you throw and catch in the same method?
    } 
}
```